ludwig/datasets/ohsumed/ohsumed.py

```diff
@@ -0,0 +1,93 @@
+#! /usr/bin/env python
+# coding=utf-8
+# Copyright (c) 2019 Uber Technologies, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+import csv
+import os
+import shutil
+from pathlib import Path
+import pandas as pd
+from ludwig.datasets.base_dataset import BaseDataset
+
+"""A class to download, process and return the ohsumed training data to be plugged into the Ludwig training API"""
+
+
+def load(cache=None):
+    dataset = OhsuMed(cache)
+    return dataset.load()
+
+
+class OhsuMed(BaseDataset):
+
+    def __init__(self, passed_in_cache_dir):
+        super().__init__("ohsumed")
+        if passed_in_cache_dir is not None:
+            self._cache_dir = passed_in_cache_dir
+        self._source_location = os.path.join(self._initial_path, "./text/ohsumed/ohsumed-allcats.csv")
+        self._cur_version = self._config_file_contents["ohsumed"]
+
+    """Download the ohsumed raw data to the ludwig cache in the format ~/.ludwig_cache/id
+       where is is represented by the name.version of the dataset
+       :param dataset_name: (str) the name of the dataset we need to retrieve.
+       Returns: 
+          None
+    """
+    def download(self, dataset_name) -> None:
+        download_path = Path.home().joinpath('.ludwig_cache').joinpath("ohsumed_"
+                                                       + str(self._cur_version)).joinpath('raw.csv')
+        shutil.copy(self._source_location, download_path)
+        result = os.path.isfile(download_path)
+        if not result:
+            raise FileNotFoundError("The raw data was not downloaded correctly")
+
+    """Process the ohsumed dataset to get it ready to be plugged into a dataframe
+       in the manner needed by the ludwig training API, to do this we create
+       a new dictionary that contains the KV pairs in the format that we need.
+       Returns:
+           None
+    """
+    def process(self) -> None:
```

Comment:
To simplify code reuse and class hierarchy, I think we should separate the public interface from the dataset-specific implementation.  Specifically, we can have the public interface implement the "high level" actions and the dataset implement only dataset-internal details:

```
# in base class
def process(self):
    if not self.downloaded():
        self.download()
    self.process_downloaded_dataset()
```

Then in the child, we implement the "protected" `process_downloaded_dataset()` function.  What do you think?

Reply:
Done, see the updates